src/main/scala/uk/gov/nationalarchives/tdr/api/auth/AuthorisationTags.scala

```diff
@@ -119,3 +119,19 @@ object ValidateUserOwnsFiles extends AuthorisationTag {
       })
   }
 }
+
+case class ValidateNoPreviousUploadForConsignment[T](argument: Argument[T]) extends AuthorisationTag {
+  override def validate(ctx: Context[ConsignmentApiContext, _])
+                       (implicit executionContext: ExecutionContext): Future[BeforeFieldResult[ConsignmentApiContext, Unit]] = {
+    val arg: T = ctx.arg[T](argument.name)
+    val consignmentId: UUID = arg match {
+      case uoc: UserOwnsConsignment => uoc.consignmentId
+      case id: UUID => id
+    }
+
+    ctx.ctx.consignmentService.previousUploadToConsignment(consignmentId).map {
+      case true => throw AuthorisationException("Upload already occurred for consignment: " + consignmentId)
```

Comment:
This is a new type of error - it's more about the consignment not being in the right state rather than an authorisation problem. (i.e. the user is allowed to upload files to this consignment - it's just the wrong stage in the transfer process for it.)

Maybe we should make a new file for tags that check consignment state? And then add a different type of exception, and a different exception representing invalid state? You can give that its own error code in the `exceptionHandler` in GraphQLServer.scala.

If possible, we should get Sangria to check the authorisation tags before the state tags, because if a user isn't allowed to do a certain action then we should tell them that rather than give them information about the consignment that they're not supposed to see.

Reply:
Thanks. I was wondering about this, as the validation checks for the consignment didn't really fit under authorisation checks.